### PR TITLE
bug: Add annotation to ModelField for use in BaseModel.__signature__

### PR DESCRIPTION
**Pull request by adriangb**
Currently, `ModelField.outer_type_` is used for `Model.__signature__`.

This leads to signatures that differ from dataclasses or classes with a manual `__init__` of the same types.

```python
from dataclasses import dataclass
from inspect import signature
from typing import Optional

from pydantic import BaseModel

@dataclass
class Dataclass:
    foo: Optional[int] = None

class CustomClass:
    def __init__(self, foo: Optional[int] = None) -> None:
        ...

class PydanticModel(BaseModel):
    foo: Optional[int] = None


print(signature(Dataclass).parameters["foo"].annotation)  # Optional[int]
print(signature(CustomClass).parameters["foo"].annotation)  # Optional[int]
print(signature(PydanticModel).parameters["foo"].annotation)  # <class 'int'>
```

This PR fixes this by storing the annotation in `ModelField.annotation` and then using that for the signature instead of `outer_type_`.
I realize this is adding a public attribute to `ModelField`, and that it is not as simple as storing the `type_` passed into the constructor (I tried to handle the case where the type is inferred from the default value, I think successfully), so this may not be the best solution, but it is the best I could come up with today.

### CODE REVIEW

**Review by samuelcolvin — CHANGES_REQUESTED** [review]
Seems good to me, but please add a change description.

I don't think this can go in v1.9.1 as it's not a bug fix for v1.9

**Comment by samuelcolvin** [comment]
please update.

**Comment by adriangb** [comment]
Hmm. While I am adding a public attribute, and that could be considered a feature, I am only doing it to fix what I would consider a bug (incorrect results from `inspect.signsture`). One of my main questions here was: can you think of a way of doing this that does not add a public attribute?

**Comment by samuelcolvin** [comment]
I don't mind the public attribute, I don't see another (simply) way.

Which of my points were you questioning? The current description, out which release this goes in?

**Comment by adriangb** [comment]
I was just questioning if this is a bug fix or a feature. I think the underlying problem is a bug, but if we're adding a public attribute it's more like a feature.

I'll add the change description.

File: changes/3413-fix-inspect-signature.md [hunk 1]
```diff
@@ -0,0 +1 @@
+fix mangling of inspect.signature for BaseModel
```
Comment (samuelcolvin) [inline changes/3413-fix-inspect-signature.md:1]:
```suggestion
fix mangling of `inspect.signature` for `BaseModel`
```

**Review by samuelcolvin — COMMENTED** [review]
Otherwise lgtm

**Comment by adriangb** [comment]
@samuelcolvin I guess we didn't merge this 😅

**Comment by samuelcolvin** [comment]
> @samuelcolvin I guess we didn't merge this sweat_smile

Got there in the end :sweat:.

Thank so much, as ever, _sorry for the delay_.

**Comment by JacobHayes** [comment]
@samuelcolvin looks like this one won't auto-merge because of the [old "request changes"](https://github.com/samuelcolvin/pydantic/pull/3413#pullrequestreview-929568495) - I think it'll need an approval to override.

**Review by samuelcolvin — APPROVED** [review]

**Comment by samuelcolvin** [comment]
Thanks @JacobHayes, missed that.

**Comment by JacobHayes** [comment]
Yup, I missed the auto-merge announcement so excited to find it in the wild!

**Comment by samuelcolvin** [comment]
Ye, it's been a god sent for this work.

**Pull request merged**